src/python/pants/backend/javascript/subsystems/nodejs.py

```diff
@@ -161,41 +241,180 @@ def to_env_dict(self) -> dict[str, str]:
+    asdf_local_tool_paths = asdf_result.local_tool_paths
+    special_strings: dict[str, Iterable[str]] = {
+        AsdfPathString.STANDARD: asdf_standard_tool_paths,
+        AsdfPathString.LOCAL: asdf_local_tool_paths,
+    }
+    nvm_dir = await _get_nvm_root()
+    expanded: list[str] = []
+    nvm_path_results = await MultiGet(
+        Get(
+            VersionManagerSearchPaths,
+            VersionManagerSearchPathsRequest(
+                env_tgt,
+                nvm_dir,
+                "versions/node",
+                f"[{NodeJS.options_scope}].search_path",
+                (".nvmrc",),
+                s if s == "<NVM_LOCAL>" else None,
+            ),
+        )
+        for s in paths
+        if s == "<NVM>" or s == "<NVM_LOCAL>"
+    )
+    for nvm_path in FrozenOrderedSet(itertools.chain.from_iterable(nvm_path_results)):
+        expanded.append(nvm_path)
+    for s in paths:
+        if s == "<PATH>":
+            expanded.extend(await Get(PathEnvironmentVariable, {}))  # noqa: PNT30: Linear search
+        elif s in special_strings:
+            expanded.extend(special_strings[s])
+        elif s == "<NVM>" or s == "<NVM_LOCAL>":
+            continue
+        else:
+            expanded.append(s)
+    return tuple(expanded)
+
+
+@rule
+async def nodejs_bootstrap(nodejs_env_aware: NodeJS.EnvironmentAware) -> NodeJsBootstrap:
+    search_paths = await Get(
+        ValidatedSearchPaths,
+        ValidateSearchPathsRequest(
+            env_tgt=nodejs_env_aware.env_tgt,
+            search_paths=tuple(nodejs_env_aware.search_path),
+            option_origin=f"[{NodeJS.options_scope}].search_path",
+            environment_key="nodejs_search_path",
+            is_default=nodejs_env_aware._is_default("search_path"),
+            local_only=FrozenOrderedSet(
+                (AsdfPathString.STANDARD, AsdfPathString.LOCAL, "<NVM>", "<NVM_LOCAL>")
+            ),
+        ),
+    )
+
+    expanded_paths = await _nodejs_search_paths(nodejs_env_aware.env_tgt, search_paths)
+
+    return NodeJsBootstrap(nodejs_search_paths=expanded_paths)
+
+
+class _BinaryPathsPerVersion(FrozenDict[str, Sequence[BinaryPath]]):
+    pass
+
+
+@rule(level=LogLevel.DEBUG, desc="Testing for Node.js binaries.")
+async def get_valid_nodejs_paths_by_version(bootstrap: NodeJsBootstrap) -> _BinaryPathsPerVersion:
+    paths = await Get(
+        BinaryPaths,
+        BinaryPathRequest(
+            search_path=bootstrap.nodejs_search_paths,
+            binary_name="node",
+            test=BinaryPathTest(
+                ["--version"], fingerprint_stdout=False
+            ),  # Hack to retain version info
+        ),
+    )
+
+    group_by_version = groupby((path for path in paths.paths), key=lambda path: path.fingerprint)
+    return _BinaryPathsPerVersion({version: tuple(paths) for version, paths in group_by_version})
+
+
+@rule(level=LogLevel.INFO, desc="Finding Node.js distribution binaries.")
```

Comment:
`INFO` level messages will actually be logged (while `DEBUG` only renders in the dynamic UI). While that should only be once per `pantsd` instance, it's probably not worth logging.